ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs

```diff
@@ -561,8 +579,128 @@ $peer-churn-governor
+  fuzzyDelay inRng (endTs0 `diffTime` startTs0) >>= go
+
+  where
+
+    -- TODO: #3396 revisit the policy for genesis 
+    increaseActivePeers :: STM m ()
+    increaseActivePeers =  do
+        mode <- getFetchMode
+        modifyTVar peerSelectionVar (\targets -> targets {
+          targetNumberOfActivePeers =
+              case mode of
+                   FetchModeDeadline ->
+                       targetNumberOfActivePeers base
+                   FetchModeBulkSync ->
+                       min 2 (targetNumberOfActivePeers base)
+        })
+
+    decreaseActivePeers :: STM m ()
+    decreaseActivePeers =  do
+        mode <- getFetchMode
+        modifyTVar peerSelectionVar (\targets -> targets {
+          targetNumberOfActivePeers =
+              case mode of
+                   FetchModeDeadline ->
+                       decrease $ targetNumberOfActivePeers base
+                   FetchModeBulkSync ->
+                       min 1 (targetNumberOfActivePeers base - 1)
+        })
+
+
+    go :: StdGen -> m Void
+    go rng = do
+      startTs <- getMonotonicTime
+
+      -- Purge the worst active peer(s).
+      atomically decreaseActivePeers
+
+      -- Short delay, we may have no active peers right now
+      threadDelay 1
+
+      -- Pick new active peer(s) based on the best performing established
+      -- peers.
+      atomically increaseActivePeers
+
+      -- Give the promotion process time to start
+      threadDelay 1
+
+      -- Forget the worst performing non-active peers.
+      atomically $ modifyTVar peerSelectionVar (\targets -> targets {
+          targetNumberOfRootPeers = decrease (targetNumberOfRootPeers base)
+        , targetNumberOfKnownPeers = decrease (targetNumberOfKnownPeers base)
+        , targetNumberOfEstablishedPeers =
+              decrease (targetNumberOfEstablishedPeers base)
+        })
+
+      -- Give the governor time to properly demote them.
+      threadDelay $ 1 + closeConnectionTimeout
+
+      -- Pick new non-active peers
+      atomically $ modifyTVar peerSelectionVar (\targets -> targets {
+          targetNumberOfRootPeers = targetNumberOfRootPeers base
+        , targetNumberOfKnownPeers = targetNumberOfKnownPeers base
+        , targetNumberOfEstablishedPeers = targetNumberOfEstablishedPeers base
+        })
+      endTs <- getMonotonicTime
+
+      fuzzyDelay rng (endTs `diffTime` startTs) >>= go
+
+    -- Randomly delay between churnInterval and churnInterval + maxFuzz seconds.
+    fuzzyDelay :: StdGen -> DiffTime -> m StdGen
+    fuzzyDelay rng execTime = do
+      mode <- atomically getFetchMode
+      case mode of
+           FetchModeDeadline -> longDelay rng execTime
+           FetchModeBulkSync -> shortDelay rng execTime
+
+    fuzzyDelay' :: DiffTime -> Double -> StdGen -> DiffTime -> m StdGen
+    fuzzyDelay' baseDelay maxFuzz rng execTime = do
+      let (fuzz, rng') = randomR (0, maxFuzz :: Double) rng
+          !delay = realToFrac fuzz + baseDelay - execTime
```

Comment:
Two things that I changed here:

* `maxFuxx` does not need a type annotation, it is specified by the type signature already
* there's no need to force `delay` thunk, it will be forced anyway by `threadDelay` just below.

Reply:
done